README.md
# flutterkhaltipayment
[![Pub](https://img.shields.io/pub/v/flutterkhaltipayment)](https://pub.dev/packages/flutterkhaltipayment)

An unofficial Khalti SDK that can be used by Khalti merchants to integrate Khalti payment on their flutter apps.

> Supported Platforms
> - Android
> - iOS (Soon)

## Usage

```yaml
# add this line to your dependencies
dependencies:
  flutterkhaltipayment: ^0.4.0
```

```dart
import 'package:flutterkhaltipayment/flutterkhaltipayment.dart';  
```

```dart
    FlutterKhaltiPayment(
      urlSchemeIOS: "KhaltiPayFlutterExampleScheme",
      merchantKey: "YOUR MERCHANT KEY HERE",
      productId: "0123",
      productName: "Product Name",
      amount: 1000,
      enableEBanking: false,
      enableIPS: false,
      enableMobileBanking: false,
      enableSCT: false,
      productUrl: "http://www.example.com/"
    ).initPayment(
      onSuccess: (data) {
        print("Token Got: ${data["token"].toString()}");
        print("Success got: ${data.toString()}");
      },
      onError: (error) {
        print("error");
        print(error);
      },
    );
```

| property           | description                                              |
| -------------------|----------------------------------------------------------|
| merchantKey        | String (Not Null)(required)                              |
| productId          | String (Not Null) (required)                             |
| productName        | String (Not Null) (required)                             |
| amount             | double (Not Null) (required)                             |
| enableEBanking     | boolean (optional)                                       |
| enableIPS          | boolean (optional)                                       |
| enableMobileBanking| boolean (optional)                                       |
| enableSCT          | boolean (optional)                                       |
| productUrl         | String (Nullable) (optional)                             |

## If you need any features suggest


CHANGELOG.md
## 0.1.0
- Initial Release.
## 0.2.0
- Added Docs
## 0.3.0
- Fixed bugs occurred during khalti payment success
## 0.4.0
- Temporarily removed feature to pass custom data while payment cause of issues
- Added Option to choose Khalti payments Types (EBanking, SCT, Connect-IPS,Mobile Banking)


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutterkhaltipayment_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


lib/flutterkhaltipayment.dart

import 'package:flutter/services.dart';
import 'package:flutter/widgets.dart';

class FlutterKhaltiPayment {
  static const MethodChannel _channel =
      const MethodChannel('bksapps/flutterkhaltipayment');
  String _merchantKey, _productId, _productName, _productUrl, _urlSchemeIOS;
  double _amount;
  bool _enableEBanking, _enableMobileBanking, _enableIPS, _enableSCT;


  FlutterKhaltiPayment({
    @required String merchantKey,
    @required String productId,
    @required String productName,
    String productUrl="",
    @required double amount,
    @required String urlSchemeIOS,
    bool enableEBanking=false,
    bool enableMobileBanking=false,
    bool enableIPS=false,
    bool enableSCT=false

  }):this._merchantKey=merchantKey,
        this._productId=productId,
        this._productName=productName,
        this._productUrl=productUrl,
        this._amount=amount,
        this._urlSchemeIOS = urlSchemeIOS,
        this._enableEBanking = enableEBanking,
        this._enableMobileBanking = enableMobileBanking,
        this._enableIPS = enableIPS,
        this._enableSCT=enableSCT;

  initPayment({Function onSuccess, Function onError}){
    final Map<String, dynamic> params = <String, dynamic>{
      "merchantKey": _merchantKey,
      "productId": _productId,
      "productName": _productName,
      "productUrl": _productUrl,
      "urlSchemeIOS": _urlSchemeIOS,
      "amount": _amount,
      "enableEbanking": _enableEBanking,
      "enableMobileBanking": _enableMobileBanking,
      "enableIPS": _enableIPS,
      "enableSCT": _enableSCT,
    };
    _channel.invokeMethod("khaltiPaymentStart", params);
    _listenToResponse(onSuccess, onError);
  }

  _listenToResponse(Function onSuccess, Function onError) {
    _channel.setMethodCallHandler((call) {
      switch (call.method) {
        case "khaltiPaymentSuccess":
          return onSuccess(call.arguments);
          break;
        case "khaltiPaymentError":
          return onError(call.arguments);
          break;
        default:
          return null;
      }
    });
  }
}


android/src/main/java/com/bksapps/flutterkhaltipayment/FlutterkhaltipaymentPlugin.java
package com.bksapps.flutterkhaltipayment;

import android.app.Activity;
import android.util.Log;

import androidx.annotation.NonNull;
import com.khalti.checkout.helper.Config;
import com.khalti.checkout.helper.KhaltiCheckOut;
import com.khalti.checkout.helper.OnCheckOutListener;
import com.khalti.checkout.helper.PaymentPreference;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

import static android.content.ContentValues.TAG;

/** FlutterkhaltipaymentPlugin */
public class FlutterkhaltipaymentPlugin implements MethodCallHandler, FlutterPlugin, ActivityAware  {
  private MethodChannel channel;
  private static Activity mContext;

  public static void registerWith(Registrar registrar) {
    mContext = registrar.activity();
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "bksapps/flutterkhaltipayment");
    channel.setMethodCallHandler(new FlutterkhaltipaymentPlugin());
  }

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getFlutterEngine().getDartExecutor(), "bksapps/flutterkhaltipayment");
    channel.setMethodCallHandler(this);
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals("khaltiPaymentStart")) {
      initKhaltiPayment(call);
    } else {
      result.notImplemented();
    }
  }

  private void initKhaltiPayment(MethodCall paymentData){
    Map<String, Object> map = new HashMap<>();

    String productId=paymentData.argument("productId").toString();
    String merchantKey=paymentData.argument("merchantKey").toString();
    String productName=paymentData.argument("productName").toString();
    String productUrl=paymentData.argument("productUrl").toString();
    double amount=paymentData.argument("amount");
    map.put("productId", productId);
    map.put("productName", productName);

    Config.Builder builder = new Config.Builder(merchantKey, productId, productName, (long)amount, new OnCheckOutListener() {
      @Override
      public void onSuccess(@NonNull Map<String, Object> data) {
        //{"idx":"TsHxaYRrDd6RBJsqtJnQTT","token":"AX7yDXcJmDYheiaQYTcDRk","amount":1,"mobile":"98XXXXXXXX","product_identity":"123","product_name":"Name"}
        channel.invokeMethod("khaltiPaymentSuccess", data);
      }

      @Override
      public void onError(@NonNull String action, @NonNull Map<String, String> errorMap) {
        Map <String, String> errorMessage= new HashMap<String, String>();
        errorMessage.put("action",action);
        errorMessage.put("message",errorMap.toString());
        channel.invokeMethod("khaltiPaymentError", errorMessage);
      }
    }).additionalData(map).paymentPreferences(new ArrayList<PaymentPreference>() {{
      add(PaymentPreference.KHALTI);
      if(paymentData.argument("enableEbanking").toString().equals("true")){
        add(PaymentPreference.EBANKING);
      }
      if(paymentData.argument("enableMobileBanking").toString().equals("true")){
        add(PaymentPreference.MOBILE_BANKING);
      }
      if(paymentData.argument("enableIPS").toString().equals("true")){
        add(PaymentPreference.CONNECT_IPS);
      }
      if(paymentData.argument("enableSCT").toString().equals("true")){
        add(PaymentPreference.SCT);
      }
    }}).productUrl(productUrl);

    KhaltiCheckOut khaltiCheckOut = new KhaltiCheckOut(mContext, builder.build());
    mContext.runOnUiThread(new Runnable() {
      public void run() {
        khaltiCheckOut.show();
      }
    });
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }

  @Override
  public void onAttachedToActivity(ActivityPluginBinding binding) {
    mContext= binding.getActivity();
  }

  @Override
  public void onDetachedFromActivityForConfigChanges() {

  }

  @Override
  public void onReattachedToActivityForConfigChanges(ActivityPluginBinding binding) {
  }

  @Override
  public void onDetachedFromActivity() {

  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutterkhaltipayment/flutterkhaltipayment.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
   @override
  void initState() {
    super.initState();
  }
  _payViaKhalti() {
    FlutterKhaltiPayment(
      urlSchemeIOS: "KhaltiPayFlutterExampleScheme",
      merchantKey: "YOUR MERCHANT KEY HERE",
      productId: "0123",
      productName: "Product Name",
      amount: 1000,
      enableEBanking: false,
      enableIPS: false,
      enableMobileBanking: false,
      enableSCT: false,
      productUrl: "http://www.example.com/"
    ).initPayment(
      onSuccess: (data) {
        print("Token Got: ${data["token"].toString()}");
        print("Success got: ${data.toString()}");
      },
      onError: (error) {
        print("error");
        print(error);
      },
    );
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: FlatButton(
            color: Colors.purple,
            child: Text("Pay via khalti", style: TextStyle(color: Colors.white),),
            onPressed: () {
              _payViaKhalti();
            },
          ),
        ),
      ),
    );
  }
}
